src/burnchains/db.rs

```diff
@@ -147,16 +262,554 @@ impl<'a> BurnchainDBTransaction<'a> {
             &u64_to_sql(header.timestamp)?,
         ];
-
         match self.sql_tx.execute(sql, args) {
             Ok(_) => Ok(self.sql_tx.last_insert_rowid()),
             Err(e) => Err(BurnchainError::from(e)),
         }
     }
 
-    fn store_blockstack_ops(
+    fn insert_block_commit_affirmation_map(
+        &self,
+        affirmation_map: &AffirmationMap,
+    ) -> Result<u64, DBError> {
+        let weight = affirmation_map.weight();
+        let sql = "INSERT INTO affirmation_maps (affirmation_map,weight) VALUES (?1,?2)";
+        let args: &[&dyn ToSql] = &[&affirmation_map.encode(), &u64_to_sql(weight)?];
+        match self.sql_tx.execute(sql, args) {
+            Ok(_) => {
+                let am_id = BurnchainDB::get_affirmation_map_id(&self.sql_tx, &affirmation_map)?
+                    .expect("BUG: no affirmation ID for affirmation map we just inserted");
+                Ok(am_id)
+            }
+            Err(e) => Err(DBError::SqliteError(e)),
+        }
+    }
+
+    fn update_block_commit_affirmation(
+        &self,
+        block_commit: &LeaderBlockCommitOp,
+        anchor_block_descendant: Option<u64>,
+        affirmation_id: u64,
+    ) -> Result<(), DBError> {
+        let sql = "UPDATE block_commit_metadata SET affirmation_id = ?1, anchor_block_descendant = ?2 WHERE burn_block_hash = ?3 AND txid = ?4";
+        let args: &[&dyn ToSql] = &[
+            &u64_to_sql(affirmation_id)?,
+            &u64_to_sql(anchor_block_descendant.unwrap_or(NO_ANCHOR_BLOCK))?,
+            &block_commit.burn_header_hash,
+            &block_commit.txid,
+        ];
+        match self.sql_tx.execute(sql, args) {
+            Ok(_) => {
+                test_debug!("Set affirmation map ID of {} - {},{},{} (parent {},{}) to {} (anchor block descendant? {:?})",
+                            &block_commit.burn_header_hash, &block_commit.txid, block_commit.block_height, block_commit.vtxindex, block_commit.parent_block_ptr, block_commit.parent_vtxindex, affirmation_id, &anchor_block_descendant);
+                Ok(())
+            }
+            Err(e) => Err(DBError::SqliteError(e)),
+        }
+    }
+
+    pub fn set_anchor_block(
+        &self,
+        block_commit: &LeaderBlockCommitOp,
+        target_reward_cycle: u64,
+    ) -> Result<(), DBError> {
+        let sql = "INSERT OR REPLACE INTO anchor_blocks (reward_cycle) VALUES (?1)";
+        let args: &[&dyn ToSql] = &[&u64_to_sql(target_reward_cycle)?];
+        self.sql_tx
+            .execute(sql, args)
+            .map_err(|e| DBError::SqliteError(e))?;
+
+        let sql = "UPDATE block_commit_metadata SET anchor_block = ?1 WHERE burn_block_hash = ?2 AND txid = ?3";
+        let args: &[&dyn ToSql] = &[
+            &u64_to_sql(target_reward_cycle)?,
+            &block_commit.burn_header_hash,
+            &block_commit.txid,
+        ];
+        match self.sql_tx.execute(sql, args) {
+            Ok(_) => {
+                test_debug!(
```

Comment:
It seems like this debug line would be useful in production -- maybe even as `info` level.

Reply:
Done